Feat: Support URL parameters to change Dashboard filters

NihadBadalov opened the pull request:
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [X] I have read and understand the pull request rules.

# Description
Support URL parameters to change Dashboard filters.

The following work:
- `/dashboard?filter=true&status=up`
- `/dashboard?filter=true&status=up&status=down`
- `/dashboard?filter=true&status=up&status=down&status=pending&status=maintenance`

If the `filter` URL parameter is set to false, the filters are not going to be in effect.

Fixes #3672 

## Type of change

Please delete any options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I ran ESLint and other linters for modified files
- [X] I have performed a self-review of my own code and tested it
- [X] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [X] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

<img width="409" alt="image" src="https://github.com/louislam/uptime-kuma/assets/32594553/4a1d1d32-1fca-4135-b209-1efe057a51c9">

<img width="440" alt="image" src="https://github.com/louislam/uptime-kuma/assets/32594553/0ab604e7-a7a9-4ab9-99ef-8e473f467195">

CommanderStorm reviewed (changes requested):
I have a few requests:
- you currently have a paramerter to enable/disable filters via the query => this is not nessesary
- Please use the vue-router's query attributes instead of `URL` on mount: https://vueschool.io/lessons/router-query-params
- It might be better to look at if storing the filters only in the url might be benefitial
- you currently have only the status being filtereable, please include the other filters too
  ![image](https://github.com/louislam/uptime-kuma/assets/26258709/5c69b271-1e24-4d75-8578-35923bbc38b2)

NihadBadalov commented:
@CommanderStorm Could you clarify what you mean by this?
> you currently have a paramerter to enable/disable filters via the query => this is not nessesary

CommanderStorm commented:
I meant that the `?filter=true` part of the URL is not necessary and should be removed.
Since no filters in the URL exist otherwise (⇒ no URL-confusion possible) and the UI has no state of “enabling filters” I think this does not need to be added.

IDK, why this was in the original issue…

NihadBadalov commented:
Got it

NihadBadalov commented:
Added URL filtering to the remaining filters.
Removed the required `filter=true` parameter.
Using vue-router instead of `URL`.

Note that tags can be searched by their names and IDs. This means the following - considering the tag `never-maint` has an ID of 1, `maint` - 2 - are equal:
- `/dashboard/3?status=up&active=running,paused&tags=never-maint,maint`
- `/dashboard/3?status=up&active=running,paused&tags=1,2`

NihadBadalov commented:
@CommanderStorm Not sure I'll be able to do this but I'll give it a try anyway. How do you want this implemented? Everything gets read off the URL queries/parameters or the information gets still stored inside of a variable (_filterState_) but all the changes get reflected on the URL?
> It might be better to look at if storing the filters only in the url might be benefitial

CommanderStorm commented:
The current approach is run via the `mounted`-lifecycle hook.
While this may work, this is a bit cumbersome and also not [reactive](https://vuejs.org/guide/essentials/reactivity-fundamentals.html).
The current approach also has the downside (I think, would need to test) of slowing down site-loads => making it feel slower.

I would shift the whole state we store for the filters to the URL.
Currently, this state is store here:
https://github.com/louislam/uptime-kuma/blob/cf2d603e277d18e96906ac36743ec4e7d58b8aad/src/components/MonitorList.vue#L98-L102

NihadBadalov commented:
Alright. Will give it a try

CommanderStorm commented on src/components/MonitorList.vue:
```diff
@@ -204,8 +215,45 @@ export default {
         },
     },
-    mounted() {
+    async mounted() {
```
This seems a bit overcomplicated. 

If all relevant state is stored in the url, why is this necessary?

NihadBadalov replied:
If you mean the variables like `activeStates` and `statusStates`, they are used for an indirection, since names are used instead of IDs

NihadBadalov replied:
Yeah `async` was unnecessary right there

CommanderStorm commented on src/components/MonitorList.vue:
```diff
@@ -95,11 +95,22 @@ export default {
-                active: null,
-                tags: null,
-            }
+            statusStates: {
```
I think you added this as a sort of indirecton between the query parameters and the state.
I think you are overcomplicating this a bit, right?

PS: In any case, since this is not mutable state (but rather a constant LUT) I think this should not live here ^^

NihadBadalov replied:
Yes. The reason I did that is so URLs like `/dashboard?status=up` can be used - which is quite nice for user experience, like the use cases of guys at #3672. This means that IDs are not used in the URLs - the names are - so, the access is done through the names.

Do you want me to move it outside and make it a const?

CommanderStorm replied:
Using ids in the url is fine by me if the benefit is having less confusing code.

I think in this case maintainability outweighs having "prettier" url-query parameters => lets keep it simple

CommanderStorm commented on src/components/MonitorListFilter.vue:
```diff
@@ -162,23 +162,26 @@ export default {
             let num = 0;
 
-            Object.values(this.filterState).forEach(item => {
+            Object.values(
```
I am unclear on why you did you add this filtering step

NihadBadalov replied:
Because the filter state was kept in `filterState`. Now it is kept in the URL, so we need to parse the URL query and filter the queries that are used for filters (status, active, tags)

CommanderStorm replied:
Okay, lets simplify this to

```js
return this.$router.currentRoute.value.query.status?.length || 0 +
    this.$router.currentRoute.value.query.active?.length || 0 +
    this.$router.currentRoute.value.query.tags?.length || 0;
    

```

CommanderStorm commented on src/components/MonitorListFilter.vue:
```diff
@@ -191,52 +194,63 @@ export default {
             } else {
-                if (newFilter.tags.includes(tag.id)) {
-                    newFilter.tags = newFilter.tags.filter(item => item !== tag.id);
-                } else {
-                    newFilter.tags.push(tag.id);
-                }
+                newFilter.tags.push(tag.name);
```
This looks like it changes the filtering logic. Is ths change correct?

NihadBadalov replied:
Yes. The change is correct and I tested it. I am converting the boolean things to string (e.g. ''+true === 'true') when comparing them because the `active` value in the URL is a "stringified boolean".

Although it works, stringifying values shouldn't have been the case for `toggleTagFilter` and `toggleStatusFilter` - only needed it for `toggleActiveFilter` to compare the passed values with the values in the URL. Thanks for the catch, will fix it in the next commit.

CommanderStorm commented on src/components/MonitorListFilter.vue:
```diff
@@ -191,52 +194,63 @@ export default {
+            );
+
+            return {
+                status: filters["status"] ? filters["status"].split(",") : [],
```
Here you try to store the query as a comma separated string.
I think there is a simpler way, as vue-router seems to allow storing lists ⇒ should be able to retrieve it
https://stackoverflow.com/questions/50692081/vue-js-router-query-array

Is there a reason for this approach I am not seeing?

NihadBadalov replied:
Didn't know that existed. Very nice, will rewrite that part for tags.

NihadBadalov replied:
@CommanderStorm Do you want me to rewrite that for every filter? Is the current approach not fine?

NihadBadalov replied:
> Here you try to store the query as a comma separated string. I think there is a simpler way, as vue-router seems to allow storing lists ⇒ should be able to retrieve it https://stackoverflow.com/questions/50692081/vue-js-router-query-array
> 
> Is there a reason for this approach I am not seeing?

@CommanderStorm Do you want me to rewrite that for every filter? The current approach works well, is it not fine?

NihadBadalov replied:
@CommanderStorm Do you want me to rewrite that for every filter? The current approach works well, is it not fine?

CommanderStorm replied:
Yes, please consider using the tools provided via the vue router instead of inventing a new way of storing lists.
This should significantly clean up the code ^^

CommanderStorm commented on src/components/MonitorListFilter.vue:
```diff
@@ -191,52 +194,63 @@ export default {
     },
     methods: {
+        getActiveFilters: function () {
+            const filters = Object.fromEntries(
```
Not a js-magician on my side.
What is happeing here?

NihadBadalov replied:
JS magic. Yeah, should've been:
```js
const filters = this.$router.currentRoute.value.queey
```

CommanderStorm reviewed (commented):
I have left a few comments where I think this PR can be simplified

NihadBadalov commented:
Will get this done sometime this week

CommanderStorm commented on src/lang/ru-RU.json:
```diff
@@ -161,6 +161,7 @@
     "Add New below or Select...": "Добавить новый или выбрать…",
     "Tag with this name already exist.": "Такой тег уже существует.",
     "Tag with this value already exist.": "Тег с таким значением уже существует.",
+    "Tag name contains a comma.": "Название тега содержит запятую.",
```
See our contribution guide: doing this might causes conflicts with Weblate

```suggestion
```

CommanderStorm reviewed (changes requested):
I don't think, in the current state, another review makes sense.

Please refer to my review above. There are a lot of them which are not resolved 
=> let's continue the discussion there.
